src/schema/date_time.rs

```diff
@@ -0,0 +1,350 @@
+            .finish()
+    }
+}
+
+/// DateTime Precision
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+#[repr(u8)]
+pub enum DateTimePrecision {
+    /// Seconds precision
+    Seconds = 0,
+    /// Milli-seconds precision.
+    Milliseconds = 1,
+    /// Micro-seconds precision.
+    Microseconds = 2,
+    /// Nano-seconds precision.
+    Nanoseconds = 3,
+}
+
+impl Default for DateTimePrecision {
+    fn default() -> Self {
+        DateTimePrecision::Milliseconds
+    }
+}
+
+impl TryFrom<u8> for DateTimePrecision {
+    type Error = String;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(DateTimePrecision::Seconds),
+            1 => Ok(DateTimePrecision::Milliseconds),
+            2 => Ok(DateTimePrecision::Microseconds),
+            3 => Ok(DateTimePrecision::Nanoseconds),
+            any => Err(format!("Invalid precision `{}` specified.", any)),
+        }
+    }
+}
+
+impl BinarySerializable for DateTimePrecision {
+    fn serialize<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        writer.write_u8(*self as u8)
+    }
+    fn deserialize<R: Read>(reader: &mut R) -> io::Result<Self> {
+        let value_u8 = reader.read_u8()?;
+        Self::try_from(value_u8).map_err(|error| io::Error::new(io::ErrorKind::Other, error))
+    }
+}
+
+/// Convert a timestamp with precision to OffsetDateTime.
+fn precise_timestamp_to_datetime(
+    timestamp: i64,
+    precision: DateTimePrecision,
+) -> Result<OffsetDateTime, String> {
+    match precision {
+        DateTimePrecision::Seconds => OffsetDateTime::from_unix_timestamp(timestamp),
+        DateTimePrecision::Milliseconds => {
+            OffsetDateTime::from_unix_timestamp_nanos((timestamp as i128) * 1_000_000)
+        }
+        DateTimePrecision::Microseconds => {
+            OffsetDateTime::from_unix_timestamp_nanos((timestamp as i128) * 1000)
+        }
+        DateTimePrecision::Nanoseconds => {
+            OffsetDateTime::from_unix_timestamp_nanos(timestamp as i128)
+        }
+    }
+    .map_err(|error| error.to_string())
+}
+
+/// A wrapper around DateTime to denote it has custom precision.
+// It only exists to provide support the underlying fast
+// field value for both `[Value::Date]` and `[Value::DateTime]` variants.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PreciseDateTime(
+    /// The inner DateTime
+    pub DateTime,
+);
+
+impl Deref for PreciseDateTime {
+    type Target = DateTime;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for PreciseDateTime {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl fmt::Debug for PreciseDateTime {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DateTime")
+            .field("timestamp", &self.0.timestamp)
+            .field("precision", &self.0.precision)
```

Comment:
But the precision is wrong isn't it? Also we don't want to print `DateTime` here.

Reply:
Precision is always correct, whether you use `DateTime::from_unix_timestamp(u64)` or `DateTime::from_utc_with_precision(utc_dt, precison)` constructor. Maybe the names should be:
- DateTime ->  DateTimeWithSecondPrecision (old DateTime)
- PreciseDateTime ->  DateTimeWithCustomPrecision (for new fast field)